qcodes/dataset/measurements.py

```diff
@@ -730,6 +737,40 @@ def _register_arrayparameter(self,
                                  basis,
                                  paramtype)
 
+    def _register_parameter_with_setpoints(self,
+                                           parameter: ParameterWithSetpoints,
+                                           setpoints: Optional[setpoints_type],
+                                           basis: Optional[setpoints_type],
+                                           paramtype: str) -> None:
+        """
+        Register an ParameterWithSetpoints and the setpoints belonging to the
+        Parameter
+        """
+        name = str(parameter)
+        my_setpoints = list(setpoints) if setpoints else []
+        for sp in parameter.setpoints:
+            if not isinstance(sp, Parameter):
+                raise RuntimeError("The setpoints of a "
+                                   "ParameterWithSetpoints "
+                                   "must be a Parameter")
```

Comment:
I kept this strict with the intention that we can loosen this requirement later if needed, but happy to change it to _baseparameter

Reply:
aha. But then what was the idea behind having the restriction inside `ParameterWithSetpoints` be `_BaseParameter` while inside `Measurement` it is `Parameter`?

I guess the idea is that in principle `ParameterWithSetpoints` can have whatever it pleases as its setpoints, right? But for `Measurement` to "work correctly" (what does it actually mean?..) the setpoints have to be something more specific? I, frankly, cannot come up with an example of what combination of the parameter and its setpoints would break `Measurement` (unless the `expand_helper` is not used because that one expects setpoint(s) to be just simple parameters which return arrays on `.get()` with correct shapes and stuff, right?).

Reply:
So I guess it would be problematic, and I don't think we should support it, if the setpoints parameter was an old array parameter or multiparameter.

Reply:
The `setpoints` need to implement `get` as you say and have a `validator` of type `Arrays` with a defined shape

Reply:
good. let's stick to this "expectation" on setpoints in general, and the restriction in `Measurement` as well for now.